go/checks.rst

```diff
@@ -0,0 +1,302 @@
+Go build-time code analysis
+===========================
+
+.. _go_library: core.rst#go_library
+.. _analysis: tools/analysis/analysis.go
+.. _GoLibrary: providers.rst#GoLibrary
+.. _GoSource: providers.rst#GoSource
+.. _GoArchive: providers.rst#GoArchive
+.. _GoChecker: providers.rst#GoChecker
+
+.. role:: param(kbd)
+.. role:: type(emphasis)
+.. role:: value(code)
+.. |mandatory| replace:: **mandatory value**
+
+rules_go allows you to define custom source-code-level checks that are executed
+alongside the Go compiler. These checks can be used to catch bugs early in the
+development process.
+
+.. contents:: :depth: 2
+
+-----
+
+Overview
+--------
+
+Writing and registering checks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Checks are Go packages are register themselves with package `analysis`_.
+Each check is invoked once per Go package, and is provided the abstract
+syntax trees (ASTs) and type information for that package. For example:
+
+.. code:: go
+
+    package mycheck
+
+    import "github.com/bazelbuild/rules_go/go/tools/analysis"
+
+    var Analysis = &analysis.Analysis{Name: "mycheck", Run: run}
+
+    func init() {
+      analysis.Register(Analysis)
+    }
+
+    func run(p *analysis.Package) (*analysis.Result, error) {
+      // Analysis logic goes here.
+    }
+
+Each check must be built as a ``go_tool_library`` target. This rule is
+identical to `go_library`_ but is defined separately to avoid bootstrapping
+problems.
+
+.. code:: bzl
+
+    go_tool_library(
+        name = "mycheck",
+        srcs = ["mycheck.go"],
+        importpath = "mycheck",
+        deps = ["@io_bazel_rules_go//go/tools/analysis:analysis"],
+        visibility = ["//visibility:public"],
+    )
+
+    go_tool_library(
+        name = "unsafedom",
+        srcs = [
+            "check_dom.go",
+            "dom_utils.go",
+        ],
+        importpath = "unsafedom",
+        deps = ["@io_bazel_rules_go//go/tools/analysis:analysis"],
+        visibility = ["//visibility:public"],
+    )
+
+You must define a ``go_checker`` target whose ``deps`` attribute contains all
```

Comment:
A short explanation of what this rule builds would be useful.

Reply:
Done.